refactor(compiler-cli): support xi18n in ngtsc

alxhub opened the pull request:
xi18n is the operation of extracting i18n messages from templates in the
compilation. Previously, only View Engine was able to perform xi18n. This
commit implements xi18n in the Ivy compiler, and a copy of the View Engine
test for Ivy verifies that the results are identical.

alxhub committed:
refactor(compiler-cli): support xi18n in ngtsc

xi18n is the operation of extracting i18n messages from templates in the
compilation. Previously, only View Engine was able to perform xi18n. This
commit implements xi18n in the Ivy compiler, and a copy of the View Engine
test for Ivy verifies that the results are identical.

petebacondarwin commented:
@alxhub is the purpose of this PR to enable migration of G3 projects that use `ngc` to do the message extraction?

alxhub commented:
> @alxhub is the purpose of this PR to enable migration of G3 projects that use `ngc` to do the message extraction?

Correct!

alxhub committed:
refactor(compiler-cli): extract xi18n utility functions to a separate file

This commit moves some xi18n-related functions in the View Engine
ng.Program into a new file. This is necessary in order to depend on them
from the Ivy ng.Program while avoiding a cycle.

petebacondarwin commented on packages/compiler-cli/index.ts:
```diff
@@ -21,7 +21,9 @@ export * from './src/tooling';
 export {CompilerOptions as AngularCompilerOptions} from './src/transformers/api';
 
 export {ngToTsDiagnostic} from './src/transformers/util';
+
 export {NgTscPlugin} from './src/ngtsc/tsc_plugin';
 export {NgtscProgram} from './src/ngtsc/program';
 
+
```
Unnecessary changes?

petebacondarwin commented on packages/compiler-cli/test/ngtsc/xi18n_spec.ts:
```diff
@@ -0,0 +1,337 @@
+  describe('ngtsc xi18n', () => {
+    if (os === 'Windows' || platform() === 'win32') {
+      // xi18n tests are skipped on Windows as the paths in the expected message files are platform-
+      // sensitive. These tests will be deleted when xi18n is removed, so it's not a major priority
+      // to make them work with Windows.
```
What is the plan for deleting `xi18n` from the code base?

alxhub replied:
`$localize` all the way!

petebacondarwin commented on packages/compiler-cli/test/ngtsc/xi18n_spec.ts:
```diff
@@ -0,0 +1,337 @@
+      // to make them work with Windows.
+
+      // Add a single disabled test to ensure the spec file is always non-empty.
+      xit('tests are disabled on Windows', () => {});
+      return;
```
Could you just move the `if` block outside the `describe()` to avoid having to add this excluded test? E.g.

```ts
runInEachFileSystem(os => {
  if (os === 'Windows' || platform() === 'win32') return;

  describe('ngtsc xi18n', () => {
    ...
  });
});

```

alxhub replied:
Would this actually fix the problem? I thought the issue was an empty `_spec` file in general, not an empty `describe()` block.

petebacondarwin replied:
I just tried this locally by adding a `return;` statement directly after the `runInEachFileSystem(os => {` block of another file on master. It passed fine. But then I moved the `return;` inside the `describe()` block and it also passed. So I am not sure what the original problem was

petebacondarwin reviewed (approved):
LGTM! A couple of nits.

I am interested to hear what the plan is going forward for message extraction in G3.
Currently localization in G3 relies upon `goog.getMsg()` and closure compiler, right?
Outside G3, localization (and message extraction) is based on the `$localize` code.

atscott reviewed (approved):
reviewed-for: public-api

jessicajaniuk reviewed (approved):
LGTM :cookie:

Just a single error message suggestion that I don't feel strongly about.

reviewed-for: public-api

petebacondarwin reviewed (approved):
reviewed-for: public-api

alxhub commented on packages/compiler-cli/src/transformers/i18n.ts:
```diff
@@ -0,0 +1,74 @@
+      return 'xlf';
+  }
+
+  throw new Error(`Unsupported format "${formatName}"`);
```
Agreed - given that this is a simple code move, though, I'd like to keep the original unchanged.

alxhub commented:
Switching to `target: patch`, as this PR applies cleanly to the patch branch, and we want to keep the compiler in sync as much as possible.
